lib/Dialect/BaseModelica/Transforms/CallCSE.cpp

```diff
@@ -47,29 +39,209 @@ mlir::LogicalResult CallCSEPass::processModelOp(ModelOp modelOp) {
+  return root;
+}
+
+void CallCSEPass::emitCse(const int emittedCSEs,
+                          llvm::SmallVectorImpl<CallOp> &equivalenceGroup,
+                          ModelOp modelOp, DynamicOp dynamicOp,
+                          mlir::RewriterBase &rewriter) {
+  assert(!equivalenceGroup.empty() && "equivalenceGroup cannot be empty");
+  auto representative = equivalenceGroup.front();
+  const auto loc = representative.getLoc();
+
+  // Emit one variable per function result
+  llvm::SmallVector<VariableOp> cseVariables;
+  for (auto result : llvm::enumerate(representative.getResults())) {
+    rewriter.setInsertionPointToStart(modelOp.getBody());
+    // Emit cse variable
+    auto variableName = "_cse" + std::to_string(emittedCSEs) + "_" +
+                        std::to_string(result.index());
+    auto cseVariable = rewriter.create<VariableOp>(
+        loc, variableName, VariableType::wrap(result.value().getType()));
+    cseVariables.push_back(cseVariable);
+
+    // Emit driver equation
+    rewriter.setInsertionPoint(dynamicOp);
+    auto equationTemplateOp = rewriter.create<EquationTemplateOp>(loc);
+    rewriter.setInsertionPointToStart(equationTemplateOp.createBody(0));
+    auto lhsOp = rewriter.create<EquationSideOp>(
+        loc, rewriter.create<VariableGetOp>(loc, cseVariable)->getResults());
+    auto rhsOp = rewriter.create<EquationSideOp>(
+        loc,
+        cloneDefUseChain(representative, rewriter)->getResult(result.index()));
+    rewriter.create<EquationSidesOp>(loc, lhsOp, rhsOp);
+
+    // Add driver equation to dynamic operation
+    rewriter.setInsertionPointToEnd(dynamicOp.getBody());
+    rewriter.create<EquationInstanceOp>(rewriter.getUnknownLoc(),
+                                        equationTemplateOp);
+  }
+
+  // Replace calls with get(s) to CSE variable(s)
+  for (auto &callOp : equivalenceGroup) {
+    rewriter.setInsertionPoint(callOp);
+
+    llvm::SmallVector<mlir::Value> results;
+    for (auto &cseVariable : cseVariables) {
+      results.push_back(
+          rewriter.create<VariableGetOp>(loc, cseVariable).getResult());
+    }
+    rewriter.replaceOp(callOp, results);
+  }
+
+  this->replacedCalls += equivalenceGroup.size();
+  ++this->emittedCSEs;
+}
+
+mlir::LogicalResult CallCSEPass::processModelOp(ModelOp modelOp) {
+  mlir::IRRewriter rewriter(modelOp);
+
   llvm::SmallVector<CallOp> callOps;
+  collectCallOps(modelOp, callOps);
+
+  llvm::SmallVector<llvm::SmallVector<CallOp>> callEquivalenceGroups;
+  buildCallEquivalenceGroups(callOps, callEquivalenceGroups);
 
-  for (EquationTemplateOp templateOp : templateOps) {
-    templateOp->walk([&](CallOp callOp) {
-      callOps.push_back(callOp);
-    });
+  if (callEquivalenceGroups.empty()) {
+    return mlir::success();
   }
 
-  
+  rewriter.setInsertionPointToEnd(modelOp.getBody());
+  auto dynamicOp = rewriter.create<DynamicOp>(rewriter.getUnknownLoc());
+  rewriter.createBlock(&dynamicOp.getRegion());
+
+  int emittedCSEs = 0;
```

Comment:
Remove this counter, and use the symbol table to automatically change the variable name if such a name is already used. Look around in the code for the `insert` method of the `SymbolTable` class.